### ISSUE TITLE
BBB 2.3: Listen only mode (kurento) not working, ICE 1004

### ISSUE THREAD

**Issue by l3d00m** [ebbba-org/ansible-role-bigbluebutton#113]
I've set up a fresh BBB 2.3 installation on Ubuntu 18.04 with `bbb_freeswitch_ipv6: false` and `bbb_container_compat: true` based on 395585682f7b04fc3ff62ee00001b36b0276b56d

I could connect with microphone successfully and talk with other users. But connecting as "listen only" gave a ICE 1004 error. 

Running `bbb-conf --check` on that server gave the following output (*I replaced the last two ip address octets with XXX and censored the domain name*):

```
# sudo bbb-conf --check
default

BigBlueButton Server 2.3.0 (2148)
                    Kernel version: 4.15.0
                      Distribution: Ubuntu 18.04.5 LTS (64-bit)
                            Memory: 25165 MB
                         CPU cores: 8

/etc/bigbluebutton/bbb-web.properties (override for bbb-web)
/usr/share/bbb-web/WEB-INF/classes/bigbluebutton.properties (bbb-web)
       bigbluebutton.web.serverURL: https://my-domain-bbb3.de
                defaultGuestPolicy: ALWAYS_ACCEPT
                 svgImagesRequired: true

/etc/nginx/sites-available/bigbluebutton (nginx)
                       server_name: my-domain-bbb3.de
                              port: 80, [::]:8080
                              port: 443 ssl

/opt/freeswitch/etc/freeswitch/vars.xml (FreeSWITCH)
                       local_ip_v4: 178.254.XXX.XXX
                   external_rtp_ip: stun:my-domain-bbb3.de
                   external_sip_ip: stun:my-domain-bbb3.de

/opt/freeswitch/etc/freeswitch/sip_profiles/external.xml (FreeSWITCH)
                        ext-rtp-ip: $${local_ip_v4}
                        ext-sip-ip: $${local_ip_v4}
                        ws-binding: :5066
                       wss-binding: 178.254.XXX.XXX:7443

/usr/local/bigbluebutton/core/scripts/bigbluebutton.yml (record and playback)
                     playback_host: my-domain-bbb3.de
                 playback_protocol: https
                            ffmpeg: 4.2.4-1ubuntu0.1bbb2~18.04

/etc/bigbluebutton/nginx/sip.nginx (sip.nginx)
                        proxy_pass: 178.254.XXX.XXX
                          protocol: https

/usr/local/bigbluebutton/bbb-webrtc-sfu/config/default.yml (Kurento SFU)
                        kurento.ip: inet
                       kurento.url: ws://127.0.0.1:8888/kurento
                    kurento.sip_ip: inet
                    localIpAddress: inet
               recordScreenSharing: true
                     recordWebcams: true
                  codec_video_main: VP8
               codec_video_content: VP8

/usr/share/meteor/bundle/programs/server/assets/app/config/settings.yml (HTML5 client)
                             build: 1669
                        kurentoUrl: wss://my-domain-bbb3.de/bbb-webrtc-sfu
                  enableListenOnly: true
                    sipjsHackViaWs: false

/etc/bigbluebutton/turn-stun-servers.xml (STUN Server)
                              stun: my-domain-bbb3.de


# Potential problems described below

# Warning: No firewall detected.  Recommend using setting up a firewall for your server
#
#     https://docs.bigbluebutton.org/2.2/customize.html#setup-a-firewall
#

grep: /opt/freeswitch/var/log/freeswitch/freeswitch.log: No such file or directory

# Warning: The setting of 178.254.XXX.XXX for proxy_pass in
#
#    /etc/bigbluebutton/nginx/sip.nginx
#
# does not match the local IP address (inet).
# (This is OK if you've manually changed the values)

# Warning: The setting of 178.254.XXX.XXX for local_ip_v4 in
#
#    /opt/freeswitch/etc/freeswitch/vars.xml
#
# does not match the local IP address (inet).
# (This is OK if you've manually changed the values)


#
# Kurento will try to connect to inet but FreeSWITCH is listening on 178.254.XXX.XXX for port 5066
#
# To fix, run the commands
#
# sudo yq w -i /usr/local/bigbluebutton/bbb-webrtc-sfu/config/default.yml freeswitch.sip_ip 178.254.XXX.XXX
# sudo chown bigbluebutton:bigbluebutton /usr/local/bigbluebutton/bbb-webrtc-sfu/config/default.yml
#

#
# Warning: Failed to verify STUN server at my-domain-bbb3.de 3478 with command
#
#    stunclient --mode full --localport 30000 my-domain-bbb3.de 3478
#

```

So, as suggested in that output I ran the following commands:
```
sudo yq w -i /usr/local/bigbluebutton/bbb-webrtc-sfu/config/default.yml freeswitch.sip_ip 178.254.XXX.XXX
sudo chown bigbluebutton:bigbluebutton /usr/local/bigbluebutton/bbb-webrtc-sfu/config/default.yml
```

And restarted with `bbb-conf --restart`. Now the "listen only" mode works again as expected.

Since I'm too unexperienced with Ansible and especially the whole WebRTC business, I'm not sure whether this is an issue with this ansible role. But if it is, it might be a good idea to include this in the role?

**Comment by l3d00m** [comment]
Interestingly enough, I know have an idea where the mysterious `inet` is coming from:

If I run the ip detection command from the bbb-install script (here https://github.com/bigbluebutton/bbb-install/blob/80ab1082a50596f91adad30609b4caf4398ae49b/bbb-install.sh#L457), it outputs `inet`. 

For some reason (can't find it yet), this is then inserted into `/usr/local/bigbluebutton/bbb-webrtc-sfu/config/default.yml`. I assume that is not the fault of the ansible role, but rather of the installation scripts?

I now worked around it by including a post_task that replaces the wrong IP:
```
  - name: Replace wrong IP in webrtc conf
    command: yq w -i /usr/local/bigbluebutton/bbb-webrtc-sfu/config/default.yml {{ item }} {{ ansible_default_ipv4.address }}
    with_items:
      - "freeswitch.sip_ip"
      - "freeswitch.ip"
      - "kurento[0].ip"
    notify:
      - restart kurento
      - restart bbb-webrtc-sfu
      - restart freeswitch
```

**Comment by BrutalBirdie** [comment]
Hey @l3d00m
thanks for reporting this issue.

It would be best to see how you deployed the server since I could not replicate this issue with my demo Server.
The error 1004 is also very generic and I can't pinpoint the exact problem from the given information.
See:
* https://docs.bigbluebutton.org/2.2/troubleshooting.html#webrtc-errors-1001-1002
* https://sipjs.com/api/0.6.0/causes/

But I am unsure why you got this error in the first place.
Since the file that is being mentioned in the error is getting overwritten here
BigBlueButton Documentation for Override https://docs.bigbluebutton.org/dev/dev23.html#local-overrides-for-configuration-settings

https://github.com/ebbba-org/ansible-role-bigbluebutton/blob/bbb/2.3/tasks/webrtc-sfu.yml#L26-L33

with the default if not changed by you

https://github.com/ebbba-org/ansible-role-bigbluebutton/blob/bbb/2.3/defaults/main.yml#L116-L156

**Comment by l3d00m** [comment]
Hi @BrutalBirdie,

thanks for taking the time and sorry for the possibly confusing report, I wasn't quite able to comprehend what happened. But my Ansible and BBB knowledge improved quite a lot since yesterday, so I can hopefully provide better information now.

I just setup fresh server with Ubuntu 18.04 minimal (VPS R8 server from 1blue for testing) using the following playbook with the role from 934f834921fdb058128767c21d2fabba3186cfe2.

```
- name: configure bbb
  hosts: test
  remote_user: root
  vars:
    bbb_container_compat: true
    bbb_letsencrypt_email: mymail@pm.me
    bbb_secret: "{{ lookup('password', 'credentials/' + inventory_hostname + '/bbb_secret length=41') }}"
    bbb_greenlight_secret: "{{ lookup('password', 'credentials/' + inventory_hostname + '/greenlight_secret length=64') }}"
    bbb_greenlight_db_password: "{{ lookup('password', 'credentials/' + inventory_hostname + '/greenlight_db_pw length=20') }}"
    bbb_freeswitch_socket_password: "{{ lookup('password', 'credentials/' + inventory_hostname + '/freeswitch_socket length=20') }}"
    bbb_api_demos_enable: false
    bbb_disable_recordings: true
    bbb_freeswitch_ipv6: false
    bbb_coturn_enable: false
    bbb_stun_servers:
      - server: "stun.l.google.com:19302"
    bbb_turn_servers: []
    bbb_dialplan_comfort_noise: false
    bbb_monitoring_all_in_one_enable: true
    bbb_default_welcome_message_footer: ""
    bbb_default_welcome_message: "Willkommen bei <b>%%CONFNAME%%</b>!"
  roles:
    - bigbluebutton

```

and host file with `my-server.de bbb_hostname=my-server.de`. I haven't changed any other defaults and also left `bbb_turn_servers` as well as `bbb_stun_servers` on their default in earlier attempts that weren't working as well.

If I now open a new meeting, I can join successfully with microphone (which is the big difference to most existing issues) and the error just comes up as soon as I select listen only. The error is `SIP_CALL_TIMEOUT` in the browser console and doesn't neccessarily throw a ICE error.

If I understand the override correctly, not all described values are overriden, most importantly missing is `freeswitch.sip_ip`, which I have manually changed as described in my last comment (I also could also override them in the `production.yaml`, then it works as well).

Those values are populated with the wrong IP in my case, for some reason it inserts `inet` there by default. This is how the freeswitch section in `/usr/local/bigbluebutton/bbb-webrtc-sfu/config/default.yml` looks like after the playbook install for me:

```yml
freeswitch:
  ip: inet
  sip_ip: inet
  port: 5066
  esl_ip: 127.0.0.1
  esl_port: 8021
  handleExternalConnections: false
  ipClassMappings:
    local:
    private:
    public:
  esl_password: XXX
```

Output of `ifconfig venet0:0`, in case it's relevant:
```
enet0:0: flags=211<UP,BROADCAST,POINTOPOINT,RUNNING,NOARP>  mtu 1500
        inet 178.254.XX.XX  netmask 255.255.255.0  broadcast 178.254.XX.XX  destination 178.254.SAME_AS_IP
        unspec 00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00  txqueuelen 0  (UNSPEC)
```

So by now I'm pretty sure this is a BBB install issue, but I'm very confused why it hasn't been reported before. What values do appear for you for `freeswitch.sip_ip` on a fresh 18.04?

In conclusion, it might be a good idea to add the following to the defaults?
```yml
bbb_webrtc_sfu_multikurento:
    freeswitch:
        sip_ip: "{{ ansible_default_ipv4.address }}"
```
This is also proposed in the (old) [install docs](https://docs.bigbluebutton.org/2.2/install.html), if you search for `sip_ip` there.

**Comment by BrutalBirdie** [comment]
I really need to see the server setup and the whole ansible process to find out what happened in your case.
Maybe we could find a date where we could take a look?

**Comment by l3d00m** [comment]
Sure, that would be good too. I'm in no rush to finish it due to the workaround, but I'm pretty flexible every day starting Tuesday. Generally afternoon or evening would be preferred.

**Comment by BrutalBirdie** [comment]
@l3d00m since you have a `.de` domain I assume CEST Timezone?

**Comment by l3d00m** [comment]
Yes correct, I'm from Germany.

**Comment by BrutalBirdie** [comment]
I cant reproduce your error with a fresh deploy.
See here:
https://github.com/ebbba-org/ansible-role-bigbluebutton/issues/119#issuecomment-840721988

You can add me on matrix so we can chat and see what went wrong.
`@brutalbirdie:matrix.org`

**Comment by Davka** [comment]
try to set a fix ```bbb_freeswitch_socket_password``` (not using ```lookup```)

**Comment by l3d00m** [comment]
After an extensive meeting with @BrutalBirdie (thanks!!) we came to the following conclusions.

I'm using an LXC VPS from 1blue and have compat mode enabled. After a deploy on the fresh server with this ansible role, the listen only mode fails because Kurento isn't configured correctly.

Relevant config bits

```yml
# /usr/local/bigbluebutton/bbb-webrtc-sfu/config/default.yml
freeswitch:
  ip: inet
  sip_ip: inet
```

```yml
# /etc/bigbluebutton/bbb-webrtc-sfu/production.yml
freeswitch:
    esl_password: MY_PW
    # that's it, no IP or SIP ip here
```

For some reason, `inet` is used, even if I "hardcode" a socket password as proposed by @Davka. We weren't able to find the root cause for that.

For that reason we thought that it'd be a good idea to add the following to the `defaults/main.yml`
```yml
bbb_webrtc_sfu_multikurento:
    freeswitch:
        sip_ip: "{{ ansible_default_ipv4.address }}"
        ip: "{{ ansible_default_ipv4.address }}"
```
This also what happens in the BBB install script here: https://github.com/bigbluebutton/bbb-install/blob/856c5da3e62552fe218f0fb4cd23c0bd7eb65b25/bbb-install.sh#L991 if I understand the logic correctly. And it's also proposed in the [old install docs](https://docs.bigbluebutton.org/2.2/install.html).

I'll open a PR to do that.
